src/main/flight/autopilot.c

```diff
@@ -91,17 +183,175 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+            // limit sum of D and A per axis because otherwise can be too aggressive when starting at speed
+     //       const float maxDAAngle = 35.0f; // limit in degrees; arbitrary.  20 is a bit too low, allows a lot of overshoot
+     //       const float pidDA = constrainf(pidD + pidA, -maxDAAngle, maxDAAngle);
+            // note: an angle of more than 35 degrees can still be achieved as P and I grow, and on diagonal
+            // note: angle mode limit is absolute max, as set in CLI default is 60
+
+            //total vector length for DA limiting, if needed:
+            float pidDA = pidD + pidA;
+            const float pidDASquared = pidDA * pidDA;
+            float mag = sqrtf(pidDASquared + prevPidDASquared);
+            if (mag > maxDAAngle) {
+                pidDA *= (maxDAAngle / mag);
+            }
+            prevPidDASquared = pidDASquared;
+
+            // ** PID Sum **
+            efAxis->pidSum = pidP + pidI + pidDA;
+
+            // Debugs... distances in cm, angles in degrees * 10, velocities cm/2
+            if (gyroConfig()->gyro_filter_debug_axis == loopAxis) {
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 0, lrintf(distanceCm));
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 4, lrintf(pidP * 10));
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 5, lrintf(pidI * 10));
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 6, lrintf(pidD * 10));
+                DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 7, lrintf(pidA * 10));
+            }
+        }
+
+        if (posHold.sticksActive) {
+            // keep update sanity check distance while sticks are out
+            posHold.sanityCheckDistance = gpsSol.groundSpeed > 500 ? gpsSol.groundSpeed * 2.0f : 1000.0f;
+            // if a Position Hold deadband is set, and sticks are outside deadband, allow pilot control in angle mode
+            posHold.pidSumCraft[AI_ROLL] = 0.0f;
+            posHold.pidSumCraft[AI_PITCH] = 0.0f;
+        } else {
+            // ** Rotate pid Sum to quad frame of reference, into pitch and roll **
+            const float headingRads = DECIDEGREES_TO_RADIANS(attitude.values.yaw);
+            const float sinHeading = sin_approx(headingRads);
+            const float cosHeading = cos_approx(headingRads);
+            posHold.pidSumCraft[AI_ROLL] = -sinHeading * posHold.efAxis[NS].pidSum + cosHeading * posHold.efAxis[EW].pidSum;
+            posHold.pidSumCraft[AI_PITCH] = cosHeading * posHold.efAxis[NS].pidSum + sinHeading * posHold.efAxis[EW].pidSum;
+        }
+    }
+
+    // ** Final output to pid.c Angle Mode at 100Hz with primitive upsampling**
+    autopilotAngle[AI_ROLL] = pt3FilterApply(&posHold.upsample[AI_ROLL], posHold.pidSumCraft[AI_ROLL]);
+    autopilotAngle[AI_PITCH] = pt3FilterApply(&posHold.upsample[AI_PITCH], posHold.pidSumCraft[AI_PITCH]);
+    // note: upsampling should really be done in earth frame, to avoid 10Hz wobbles if pilot yaws and the controller is applying significant pitch or roll
```

Comment:
(From my experience, lag can be more dangerous than wobble - quad can be hunting target in increasing circles)

Reply:
Yes, for sure.  Smoothing does make for a smoother picture but definitely can lead to circling and vagueness.